libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -1026,6 +1026,15 @@ void GCS_MAVLINK::update_send()
     send_ftp_replies();
 #endif // HAVE_FILESYSTEM_SUPPORT
 
+#ifdef ENABLE_SCRIPTING
+    {
+        AP_Scripting *scripting = AP_Scripting::get_singleton();
+        if (scripting != nullptr) {
+            scripting->send_message(chan);
```

Comment:
The placement here is interesting.

This is taking away from the slice of time set aside for this gcs mavlink backend to send everything it needs to send, and there's not a lot of that.  Putting this here could starve all of the rest of the sending code, both of time and (particularly!) buffer space.  That could be very bad, as you could lose telemetry entirely, depending on GCS.


There are some options to avoid that.  One would be to `send_message(ap_message_t::scripting_message)` within the scripting code so that it's handled by the deferred-message code at least semi-fairly.

Given the problems that could arise if scripting goes insane on messages, it may actually be better to add an ID to the scheduled IDs instead - i.e. add it to a "stream".  This would be safest - perhaps at the expense of significantly reducing the script's ability to rapidly send messages.

Reply:
Yeah, I really struggled with what to do here. The problem with stream ID's is it's not clear how to adjust the rate up or down for that. The other current problem with that is that we are currently using a single buffer for all the channels. This saves us on memory cost, but we have to drain the buffer in order. That means that a slow channel blocks faster channels. The alternative is we could allocate a buffer on demand for each channel as we send to it, but it means a a much larger memory footprint for it.

Reply:
There is a small part of me that says this is maybe in the realm of self inflicted for a user and maybe that's tolerable... I'm not fully sure on that one.